wcfsetup/install/files/lib/acp/form/CaptchaQuestionAddForm.class.php

```diff
@@ -46,155 +37,76 @@ class CaptchaQuestionAddForm extends AbstractForm
     /**
      * @inheritDoc
      */
-    public function assignVariables()
-    {
-        parent::assignVariables();
-
-        I18nHandler::getInstance()->assignVariables();
-
-        WCF::getTPL()->assign([
-            'action' => 'add',
-            'isDisabled' => $this->isDisabled,
-            'invalidRegex' => $this->invalidRegex,
-        ]);
-    }
+    public $objectActionClass = CaptchaQuestionAction::class;
 
     /**
      * @inheritDoc
      */
-    public function readFormParameters()
-    {
-        parent::readFormParameters();
-
-        I18nHandler::getInstance()->readValues();
-
-        if (isset($_POST['isDisabled'])) {
-            $this->isDisabled = 1;
-        }
-    }
+    public $objectEditLinkController = CaptchaQuestionEditForm::class;
 
     /**
      * @inheritDoc
      */
-    public function readParameters()
+    protected function createForm()
     {
-        parent::readParameters();
-
-        I18nHandler::getInstance()->register('question');
-        I18nHandler::getInstance()->register('answers');
-    }
-
-    /**
-     * @inheritDoc
-     */
-    public function save()
-    {
-        parent::save();
-
-        $this->objectAction = new CaptchaQuestionAction([], 'create', [
-            'data' => \array_merge($this->additionalFields, [
-                'answers' => I18nHandler::getInstance()->isPlainValue('answers') ? I18nHandler::getInstance()->getValue('answers') : '',
-                'isDisabled' => $this->isDisabled,
-                'question' => I18nHandler::getInstance()->isPlainValue('question') ? I18nHandler::getInstance()->getValue('question') : '',
-            ]),
+        parent::createForm();
+
+        $this->form->appendChildren([
+            FormContainer::create('data')
+                ->appendChildren([
+                    TextFormField::create('question')
+                        ->label('wcf.acp.captcha.question.question')
+                        ->maximumLength(255)
+                        ->required()
+                        ->i18n()
+                        ->languageItemPattern('wcf.captcha.question.question.question\d+'),
+                    MultilineTextFormField::create('answers')
+                        ->label('wcf.acp.captcha.question.answers')
+                        ->required()
+                        ->i18n()
+                        ->languageItemPattern('wcf.captcha.question.answers.question\d+')
+                        ->addValidator(new FormFieldValidator('regex', function (MultilineTextFormField $formField) {
+                            if ($formField->getValidationErrors() !== []) {
+                                return;
+                            }
+
+                            // returns `true` for a valid answer
+                            $answerValidator = static function (string $text): bool {
```

Comment:
I was not a fan of the closure even in my initial review and now I'm even less of a fan. Please split this into a `private function` with an appropriate name. May I suggest: `private function validateAnswer(string $answer): bool`?